proxy/coco/src/peer/run_state.rs

```diff
@@ -0,0 +1,383 @@
+/// succeeded.
+// TODO(xla): Review duration.
+const DEFAULT_SYNC_PERIOD: Duration = Duration::from_secs(5);
+
+/// Instructions to issue side-effectful operations which are the results from state transitions.
+#[derive(Debug, PartialEq)]
+pub enum Command {
+    /// Start the announcement subroutine.
+    Announce,
+    /// Initiate a full sync with `PeerId`.
+    SyncPeer(PeerId),
+    /// Start sync timeout.
+    StartSyncTimeout(Duration),
+}
+
+/// Significant events that occur during [`Peer`] lifetime.
+#[allow(clippy::large_enum_variant)]
+#[derive(Clone, Debug)]
+pub enum Event {
+    /// Announcement subroutine lifecycle events.
+    Announce(AnnounceEvent),
+    /// Events from the underlying coco protocol.
+    Protocol(ProtocolEvent<Gossip>),
+    /// Lifecycle events during peer sync operations.
+    PeerSync(SyncEvent),
+    /// Scheduled timeouts which can occur.
+    Timeout(TimeoutEvent),
+}
+
+/// Announcement subroutine lifecycle events.
+#[derive(Clone, Debug)]
+pub enum AnnounceEvent {
+    /// Operation failed.
+    Failed,
+    /// Operation succeeded and emitted the enclosed list of updates.
+    Succeeded(announcement::Updates),
+    /// The ticker duration has elapsed.
+    Tick,
+}
+
+/// Lifecycle events during peer sync operations.
+#[derive(Clone, Debug)]
+pub enum SyncEvent {
+    /// A sync has been initiated for `PeerId`.
+    Started(PeerId),
+    /// A sync has failed for `PeerId`.
+    Failed(PeerId),
+    /// A sync has succeeded for `PeerId`.
+    Succeeded(PeerId),
+}
+
+/// Scheduled timeouts which can occur.
+#[derive(Clone, Debug)]
+pub enum TimeoutEvent {
+    /// Grace period is over signaling that we should go offline, no matter how many syncs have
+    /// succeeded.
+    SyncPeriod,
+}
+
+/// The current status of the local peer and its relation to the network.
+#[derive(Debug)]
+enum Status {
+    /// Nothing is setup, not even a socket to listen on.
+    Stopped(Instant),
+    /// Local peer is listening on a socket but has not connected to any peers yet.
+    Started(Instant),
+    /// The local peer lost its connections to all its peers.
+    Offline(Instant),
+    /// Phase where the local peer tries get up-to-date.
+    Syncing(Instant, usize),
+    /// The local peer is operational and is able to interact with the peers it has connected to.
+    Online(Instant),
+}
+
+/// Set of knobs to change the behaviour of the [`RunState`].
+pub struct Config {
+    /// Number of peers a full sync is attempting with up on startup.
+    pub sync_max_peers: usize,
+    /// Enables the syncing stage when coming online.
+    pub sync_on_startup: bool,
+    /// Duration until the local peer goes online regardless if and how many syncs have succeeded.
+    pub sync_period: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            sync_max_peers: DEFAULT_SYNC_MAX_PEERS,
+            sync_on_startup: false,
```

Comment:
I might answer it throughout the review, but what is the idea behind defaulting to no sync on startup? I could imagine that since it requires more infrastructure to be in place that it might get in the way when running tests and what not, so just leaving a note for now.

Reply:
It's quite a noisy and bandwidth intense behaviour and should be done with care. It's a stop-gap solution until more effective dissemination strategies have settled in the protocol. Where this application change will help to inform. I'd err on the side of caution here, even so if we make it the default behaviour for upstream instances for the foreseeable future.

It does help for the tests as well as it is a happening on startup, in either case it's good to have a knob. Happy to revise the default if it is a bigger surprise to see it disabled.

Reply:
Thanks for the insight! Yeah, this seems like the right thing to do then :)